### PR TITLE
feat(remap transform): add ip address remap functions

### LINKED ISSUE #1978: New `extract_subnet` transform
It would be nice to have a transform which would take a string containing an IP address (IPv4 or IPv6) and extract a specified subnet for it.

For example, if the IP is `1.2.3.4` and the `subnet` parameter is set to `/24`, then the output would be `1.2.3.0`. There could also be other ways to configure it, for example by specifying the mask like `255.255.255.0` which would be applied to the address. It is also desirable to have ability to configure these settings differently for IPv4 and IPv6.

This might be useful for anonymizing IP addresses from logs of a public webserver before they are stored anywhere, for example to avoid storing personally identifiable information about the requesters.

### LINKED ISSUE #3761: New IP address functions for the remap syntax
Working with IP addresses is a common use case when dealing with log data. We should think about functions that help users extract (#1978) parts of an IP address and [process](https://vector.dev/docs/reference/transforms/swimlanes/#field_nameip_cidr_contains). You can see how these needs are already represented in Vector's current features with the aforementioned links.

I'm leaving this issue open so that we can agree on a set of functions that would allow users to:

- [ ] Extract parts of IPv4 and IPv6 addresses.
- [ ] Check if addresses are within larger subsets of addresses ([ex](https://vector.dev/docs/reference/transforms/swimlanes/#field_nameip_cidr_contains)).

### PR DESCRIPTION
**Pull request by FungusHumungus**
Closes #3761 
Closes #1978 

This adds the functions referenced [here](https://github.com/timberio/vector/issues/3761#issuecomment-704234558).

### CODE REVIEW

File: src/remap/function/ip_subnet.rs [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+                    .type_def(state)
+                    .fallible_unless(value::Kind::Bytes),
+            )
+            .with_constraint(value::Kind::Bytes)
+    }
+}
+
+/// Parses a subnet in the form "/8" returns the number.
+fn parse_subnet(subnet: &str) -> Result<u32> {
+    let re = Regex::new(r"/(?P<subnet>\d*)").unwrap();
```
Comment (FungusHumungus) [inline src/remap/function/ip_subnet.rs:105]: I need move this regex out into a lazy static.

File: src/remap/function/ip_subnet.rs [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+            Err("attempting to mask an ipv6 address with an ipv4 mask".into())
+        }
+        (IpAddr::V4(_), IpAddr::V6(_)) => {
+            Err("attempting to mask an ipv4 address with an ipv6 mask".into())
+        }
+    }
+}
+
+/// Returns a vector with the left `subnet_bits` set to 1,
+/// The remaining are set to 0, to make up a total length of `bytes`.
```
Comment (FungusHumungus) [inline src/remap/function/ip_subnet.rs:145]: I would be very interested to know if there is a better way this function could be written.

**Comment by FungusHumungus** [comment]
Oops, just remembered I haven't done any behaviour tests yet.

File: src/remap/function/ipv6_to_ip.rs [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+                Some(addr) => Ok(Value::from(addr.to_string())),
+                None => Err(format!("IPV6 address {} is not compatible with IPV4", addr).into()),
+            },
+        }
+    }
+
+    fn type_def(&self, state: &state::Compiler) -> TypeDef {
+        self.value
+            .type_def(state)
+            .fallible_unless(value::Kind::Bytes)
```
Comment (FungusHumungus) [inline src/remap/function/ipv6_to_ip.rs:61]: @JeanMertz Is this correct? The function will error if the parameter is not a valid ip address. Does this function need to capture this?

File: docs/reference/remap/ip_to_ipv6.cue [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+			name:        "value"
+			description: "The IPv4 ip address to convert."
+			required:    true
+			type: ["string"]
+		},
+	]
+	return: ["string"]
+	category: "ip"
+	description: #"""
+		Converts an IPv4 address to an IPv4 mapped IPv6 address.
```
Comment (bruceg) [inline docs/reference/remap/ip_to_ipv6.cue:15]: Should this pass-through an IPv6 address?

File: docs/reference/remap/ipv6_to_ip.cue [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+			name:        "value"
+			description: "The IPv4 mapped IPv6 address to convert."
+			required:    true
+			type: ["string"]
+		},
+	]
+	return: ["string"]
+	category: "ip"
+	description: #"""
+		Converts an IPv4 mapped IPv6 address to an IPv4 address.
```
Comment (bruceg) [inline docs/reference/remap/ipv6_to_ip.cue:15]: Should this be named either `ipv6_to_ipv4` or shorter `ipv6_to_v4`? IPv6 addresses *are* IP addresses.

File: src/remap/function/ip_subnet.rs (hunk 2 above)
Comment (bruceg) [reply src/remap/function/ip_subnet.rs:150]: For v4 you could simply do `~0u32 << (32 - subnet_bits)`
v6 would be similar `~0u128 << (128 - subnet_bits)`
Both `Ipv4Addr` and `Ipv6Addr` support `from` on that integer type.

File: src/remap/function/ip_to_ipv6.rs [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+        let ip = {
+            let bytes = self.value.execute(state, object)?.try_bytes()?;
+            String::from_utf8_lossy(&bytes)
+                .parse()
+                .map_err(|err| format!("unable to parse IP address: {}", err))?
+        };
+
+        match ip {
+            IpAddr::V4(addr) => Ok(Value::from(addr.to_ipv6_mapped().to_string())),
+            IpAddr::V6(addr) => Ok(Value::from(addr.to_string())),
```
Comment (bruceg) [inline src/remap/function/ip_to_ipv6.rs:51]: Ah, IPv6 *is* pass-through. This should be noted in the documentation.

**Review by bruceg — CHANGES_REQUESTED** [review]

File: src/remap/function/ip_subnet.rs (hunk 2 above)
Comment (FungusHumungus) [reply src/remap/function/ip_subnet.rs:150]: @bruceg I knew you would have a better way! Thanks.

File: docs/reference/remap.cue [hunk 7]
```diff
@@ -30,7 +30,7 @@ remap: {
 
 			arguments: [...#Argument] // Allow for empty list
 			return: [#RemapReturnTypes, ...#RemapReturnTypes]
-			category:    "coerce" | "parse" | "text" | "hash" | "event"
+			category:    "coerce" | "parse" | "text" | "hash" | "event" | "ip"
```
Comment (JeanMertz) [inline docs/reference/remap.cue:33]: minor nit: do we want a whole separate category for these functions?

If so, maybe `networking` would be a more general alternative?

File: docs/reference/remap/ipv6_to_ipv4.cue [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+		This function will raise an error if the input address is not a compatible address.
+		"""#
+	examples: [
+		{
+			title: "Success"
+			input: {
+				address: "::ffff:192.168.0.1"
+			}
+			source: #"""
+				.v4 = ipv6_to_ip(.address)
```
Comment (JeanMertz) [inline docs/reference/remap/ipv6_to_ipv4.cue:25]:
```suggestion
				.v4 = ipv6_to_ipv4(.address)
```

File: docs/reference/remap/ipv6_to_ipv4.cue [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+				v4:      "192.168.0.1"
+			}
+		},
+		{
+			title: "Error"
+			input: {
+				address: "2001:0db8:85a3::8a2e:0370:7334"
+			}
+			source: #"""
+				.v4 = ipv6_to_ip(.address)
```
Comment (JeanMertz) [inline docs/reference/remap/ipv6_to_ipv4.cue:38]:
```suggestion
				.v4 = ipv6_to_ipv4(.address)
```

File: src/remap/function/ip_cidr_contains.rs [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+#[derive(Debug, Clone)]
+struct IpCidrContainsFn {
+    cidr: Box<dyn Expression>,
+    value: Box<dyn Expression>,
+}
+
+impl IpCidrContainsFn {
+    #[cfg(test)]
+    fn new(cidr: Box<dyn Expression>, value: Box<dyn Expression>) -> Self {
+        Self { value, cidr }
```
Comment (JeanMertz) [inline src/remap/function/ip_cidr_contains.rs:44]: nit

```suggestion
        Self { cidr, value }
```

File: src/remap/function/ip_subnet.rs [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+    }
+}
+
+/// Parses a subnet in the form "/8" returns the number.
+fn parse_subnet(subnet: &str) -> Result<u32> {
+    let subnet = RE
+        .captures(subnet)
+        .ok_or_else(|| format!("{} is not a valid subnet", subnet))?;
+
+    let subnet = subnet["subnet"].parse().unwrap(); // The regex ensures these are only digits.
```
Comment (JeanMertz) [inline src/remap/function/ip_subnet.rs:113]:
```suggestion
    let subnet = subnet["subnet"].parse().expect("digits ensured by regex");
```

File: src/remap/function/ip_subnet.rs [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+    }
+}
+
+/// Returns an ipv4 address that masks out the given number of bits.
+fn ipv4_mask(subnet_bits: u32) -> IpAddr {
+    let bits = !0u32 << (32 - subnet_bits);
+    Ipv4Addr::from(bits).into()
+}
+
+/// Returns an ipv4 address that masks out the given number of bits.
```
Comment (JeanMertz) [inline src/remap/function/ip_subnet.rs:153]:
```suggestion
/// Returns an ipv6 address that masks out the given number of bits.
```

**Review by JeanMertz — CHANGES_REQUESTED** [review]

**Pull request merged**